pedalboard/process.h

```diff
@@ -165,75 +262,120 @@ process<float>(const py::array_t<float, py::array::c_style> inputArray,
-          // We're de-interleaving the data here, so we can't use std::copy.
-          for (unsigned int j = blockStart; j < blockEnd; j++) {
-            ioBufferChannelPointers[i][j] =
-                static_cast<float *>(inputInfo.ptr)[j * numChannels + i];
+    // Actually run the plugins over the ioBuffer, in small chunks, to minimize
+    // memory usage:
+    int startOfOutputInBuffer = 0;
+    int lastSampleInBuffer = 0;
+
+    for (auto *plugin : plugins) {
+      if (plugin == nullptr)
+        continue;
+
+      int pluginSamplesReceived = 0;
+
+      for (unsigned int blockStart = startOfOutputInBuffer;
+           blockStart < (unsigned int)intendedOutputBufferSize;
+           blockStart += bufferSize) {
+        unsigned int blockEnd =
+            std::min(blockStart + bufferSize,
+                     static_cast<unsigned int>(intendedOutputBufferSize));
+        unsigned int blockSize = blockEnd - blockStart;
+
+        auto ioBlock = juce::dsp::AudioBlock<float>(
+            ioBuffer.getArrayOfWritePointers(), ioBuffer.getNumChannels(),
+            blockStart, blockSize);
+        juce::dsp::ProcessContextReplacing<float> context(ioBlock);
+
+        int outputSamples = plugin->process(context);
+        pluginSamplesReceived += outputSamples;
+
+        int missingSamples = blockSize - outputSamples;
+        if (missingSamples > 0 && pluginSamplesReceived > 0) {
+          // This can only happen if the plugin we're using is returning us more
+          // than one chunk of audio that's not completely full, which can
+          // happen sometimes. In this case, we would end up with gaps in the
+          // audio output:
+          //               empty  empty  full   part
+          //              [______|______|AAAAAA|__BBBB]
+          //   end of most recently rendered block-->-^
+          // We need to consolidate those gaps by moving them forward in time.
+          // To do so, we take the section from the earliest known output to the
+          // start of this block, and right-align it to the left side of the
+          // current block's content:
+          //               empty  empty  part   full
+          //              [______|______|__AAAA|AABBBB]
+          //   end of most recently rendered block-->-^
+          for (int c = 0; c < ioBuffer.getNumChannels(); c++) {
+            // Only move the samples received before this latest block was
+            // rendered, as audio is right-aligned within blocks by convention.
+            int samplesToMove = pluginSamplesReceived - outputSamples;
+            float *outputStart =
+                ioBuffer.getWritePointer(c) + totalOutputLatencySamples;
+            float *expectedOutputEnd =
+                ioBuffer.getWritePointer(c) + blockEnd - outputSamples;
+            float *expectedOutputStart = expectedOutputEnd - samplesToMove;
+
+            std::memmove((char *)expectedOutputStart, (char *)outputStart,
```

Comment:
Consider [std::reverse_copy](https://en.cppreference.com/w/cpp/algorithm/reverse_copy) or [std::copy_n](https://en.cppreference.com/w/cpp/algorithm/copy_n) with a reverser iterator. std::memmove can work with overlaps, but i think copying in reverse is safer. May also be more efficient too, but I can't confirm that. For all I know memmove already acts that way.